src/vision/image_conversion.zig

```diff
@@ -0,0 +1,1037 @@
+    const x: isize = @intFromFloat(u * @as(f32, @floatFromInt(source_image.width)));
+    const y: isize = @intFromFloat(v * @as(f32, @floatFromInt(source_image.height)));
+
+    return getPixelClamped(source_image, x, y);
+    // return source_image.pixels[y * source_image.width + x];
+}
+
+/// Linear interpolate between two values.
+fn lerp(a: f32, b: f32, t: f32) f32 {
+    return a + (b - a) * t;
+}
+
+// Bilinear interpolate between four values.
+fn bilerp(p00: f32, p10: f32, p01: f32, p11: f32, x_fractional: f32, y_fractional: f32) f32 {
+    // p00 ----🔵---- p10
+    //         |
+    //         |
+    //         |
+    // p01 ----🟠---- p11
+    //
+    // Interpolate between the first set of values in the x-direction to find the blue dot
+    const x1 = lerp(p00, p10, x_fractional);
+    // Interpolate between the second set of values in the x-direction to find the orange dot
+    const x2 = lerp(p01, p11, x_fractional);
+
+    // Interpolate between the two new intermediate values in the y-direction to find the green dot
+    //
+    // p00 ----x1---- p10
+    //         |
+    //         🟢
+    //         |
+    // p01 ----x2---- p11
+    const result = lerp(x1, x2, y_fractional);
+    return result;
+}
+
+// Reference:
+//  - https://blog.demofox.org/2015/08/15/resizing-images-with-bicubic-interpolation/
+//  - https://en.wikipedia.org/wiki/Bilinear_interpolation
+//  - https://bartwronski.com/2021/02/15/bilinear-down-upsampling-pixel-grids-and-that-half-pixel-offset/
+pub fn sampleBilinear(source_image: anytype, u: f32, v: f32) std.meta.Child(@TypeOf(source_image.pixels)) {
+    // Calculate coordinates. Offset by half a pixel so that we are measuring to/from
+    // the center of a pixel which can then can be rounded back down to the pixel
+    // coordinate. This allows sampling to occur evenly around a pixel position and
+    // integer truncation works correctly around the edges/corners.
+    const x = (u * @as(f32, @floatFromInt(source_image.width))) - 0.5;
+    const x_fractional = x - @floor(x);
+    const y = (v * @as(f32, @floatFromInt(source_image.height))) - 0.5;
+    const y_fractional = y - @floor(y);
+
+    const p00 = getPixelClamped(source_image, @intFromFloat(x + 0), @intFromFloat(y + 0));
+    const p10 = getPixelClamped(source_image, @intFromFloat(x + 1), @intFromFloat(y + 0));
+    const p01 = getPixelClamped(source_image, @intFromFloat(x + 0), @intFromFloat(y + 1));
+    const p11 = getPixelClamped(source_image, @intFromFloat(x + 1), @intFromFloat(y + 1));
+
+    // Interpolate bi-linearly!
+    const PixelType = std.meta.Child(@TypeOf(source_image.pixels));
+    switch (PixelType) {
+        RGBPixel => {
+            return .{
+                .r = bilerp(p00.r, p10.r, p01.r, p11.r, x_fractional, y_fractional),
+                .g = bilerp(p00.g, p10.g, p01.g, p11.g, x_fractional, y_fractional),
+                .b = bilerp(p00.b, p10.b, p01.b, p11.b, x_fractional, y_fractional),
+            };
+        },
```

Comment:
For some reason, this gives different results than what I see in other applications. Haven't nailed this down yet :thinking: 

### Our implementation

```
zig build test -freference-trace --verbose --summary all -Dtest-filter="resizeImage"
```

![](https://github.com/MadLittleMods/fps-aim-analyzer/assets/558581/891018e4-934c-4079-abe6-d6792cf99762)


### Benchmark

Comparing this to the `PIL.image.resize` result from https://medium.com/hackernoon/how-tensorflows-tf-image-resize-stole-60-days-of-my-life-aba5eb093f35, ours is different.
![](https://github.com/MadLittleMods/fps-aim-analyzer/assets/558581/b90d33a0-037d-48a3-9c9b-ee7b6162a51e)

Also tested this result myself:
```py
from PIL import Image

with Image.open("images/test-square.png") as im:
    new_width = 4
    new_height = 4
    im_resized = im.resize((new_width, new_height), Image.Resampling.BILINEAR)
    im.show()
    im_resized.show()
```

![PIL resized image](https://github.com/MadLittleMods/fps-aim-analyzer/assets/558581/07bca894-c3ba-48ee-8937-1a42075d4f93)

Chrome (25% zoom):

![Chrome at 25% zoom](https://github.com/MadLittleMods/fps-aim-analyzer/assets/558581/af178211-9c5f-4300-ad15-e3ece1fa7846)

Chrome (resized `<img>`):

![Chrome img tag](https://github.com/MadLittleMods/fps-aim-analyzer/assets/558581/67bac02e-fa93-4b5e-8ff7-a3cd627edbf4)



Firefox (resized `<img>`):

![Firefox img tag](https://github.com/MadLittleMods/fps-aim-analyzer/assets/558581/9b0c860a-bbbe-43e7-a0c1-a1babc75e478)



Gimp (linear resize):

![Gimp linear resize](https://github.com/MadLittleMods/fps-aim-analyzer/assets/558581/f5bd85c1-776c-405f-a370-2361efbddeff)

Reply:
I made things better by iterating the uv in the range `[0.5/width, 1 – 0.5/width]` (https://github.com/MadLittleMods/fps-aim-analyzer/pull/8/commits/1fa7179a48f2c536be15f82101a461db06c8bf37) thanks to this comment:

> If you have a 1D texture of size 4, what are your pixel/texel coordinates?
> 
> They can be [0, 1, 2, 3]. But GPUs use a convention of half pixel offsets, so they end up being [0.5, 1.5, 2.5, 3.5]. This translates to UVs, or “normalized” coordinates [0.5/4, 1.5/4, 2.5/4, 3.5/4], which spans a range of [0.5/width, 1 – 0.5/width].
> 
> This representation seems counterintuitive at first, but what it provides us is a guarantee and convention that the image corners are placed at [0 and 1] normalized, or [0, width] unnormalized.
>
> *-- https://bartwronski.com/2021/02/15/bilinear-down-upsampling-pixel-grids-and-that-half-pixel-offset/*

![](https://github.com/MadLittleMods/fps-aim-analyzer/assets/558581/56c3c3c5-37a6-440a-bc9e-a7292ed20ede)

This made things "better" but still not the same as these other applications. It's obvious to see that it isn't taking the blue into account to make purple and even from the math it's obvious. I'm not sure how these other applications ending up with that result using "standard" bilinear.

Reply:
I added a new interpolation resize method that just averages all pixels in the UV region for the new pixel `sampleAverage` (https://github.com/MadLittleMods/fps-aim-analyzer/pull/8/commits/001ca05feb3237b98111e9f1443fbf3d163031ca) which results the ~exact same image that Chrome gives. Would be good to find the code that is responsible for this behavior.

![](https://github.com/MadLittleMods/fps-aim-analyzer/assets/558581/d47019b3-74e7-4da1-a9e0-355bb5331cb8)

Reply:
The logic for the Ladybird browser (Serenity) is in [`to_gfx_scaling_mode(...)` -> `Userland/Libraries/LibWeb/CSS/ComputedValues.h#L270-L285`](https://github.com/SerenityOS/serenity/blob/99fbd33d7d887723372e861be27f2743d335ecb5/Userland/Libraries/LibWeb/CSS/ComputedValues.h#L270-L285). If the source is bigger than what you're trying to display, it does box sampling (averaging everything), otherwise bilinear